DOC: interpolate.BarycentricInterpolator: documentation improvements

j-bowhay opened the pull request:
Various doc improvements:

- More detailed notes section
- Proper reference
- Extra example on Chebyshev points

j-bowhay committed:
DOC: interpolate.BarycentricInterpolator: documentation improvements

[docs only]

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -538,20 +538,13 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
     efficient changing of the y-values to be interpolated,
     and updating by adding more x- and y-values.
 
-    For reasons of numerical stability, this function does not compute
```
Removed as this seems too vague to be useful

ev-br replied:
Was a nice factoid IMO. Fine to remove though if you feel like it's extraneous.

j-bowhay replied:
I'll move to the notes since it is an interesting point; it just didn't seem to be necessary to include it in the summary

j-bowhay commented:
More doc build locally was behaving a bit funny, let's see how this turns out in CI

j-bowhay committed:
Apply suggestions from code review

[docs only]

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -562,7 +555,8 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
         This allows for the reuse of the weights `wi` if several interpolants
-        are being calculated using the same nodes `xi`, without re-computation.
+        are being calculated using the same nodes `xi`, without re-computation. This
+        also allows for computing the weights explicitly some choices of `xi` (see notes).
```
missing "for" in "for some choices of `xi`"?

mdhaber replied:
Is this missing a word, maybe "for"?

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+    .. math::
+
+        p(x) =
+        \frac{\sum_{i=1}^m\ w_i y_i / (x - x_i)}{\sum_{j=1}^m w_i / (x - x_i)},
```
```suggestion
        \frac{\sum_{i=1}^m\ w_i y_i / (x - x_i)}{\sum_{j=1}^m w_j / (x - x_j)},
```

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+        p(x) =
+        \frac{\sum_{i=1}^m\ w_i y_i / (x - x_i)}{\sum_{j=1}^m w_i / (x - x_i)},
+
+    where :math:`w_i` are the barycentric weights computed with to the general
```
```suggestion
    where :math:`w_i` are the barycentric weights computed with the general
```

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+    polynomial interpolation caused by floating-point arithmetic. However, it does not
+    avoid issues that are intrinsic to polynomial interpolation. Namely, if the
+    x-coordinates `xi` are equally spaced then the weights `wi` can be computed
+    explicitly as
```
Now taht you're adding references, maybe add one here for this fact, if handy (presumingly, it's [2]?)

mdhaber replied:
Looks like this was addressed.

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
-    Based on Berrut and Trefethen 2004, "Barycentric Lagrange Interpolation".
+    .. math::
+
+        w_i = (-1)^i C^n_i,
```
where $C^n_i$ is the binomial coefficient? Then is it n-choose-i or i-choose-n? --- to my eyes and in my russian textbooks it's $C_n^i = n! / i! (n-i)!$ --- the power in the binomial expansion of $(1 + x)^n$ is the lower index to $C_n^k$, unlike in the $\left(  n, k \right)$  notation. Either way is fine, just wanted to double-check :-).

mdhaber commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+    The barycentric representation avoids many of the problems associated with
+    polynomial interpolation caused by floating-point arithmetic. However, it does not
+    avoid issues that are intrinsic to polynomial interpolation. Namely, if the
+    x-coordinates `xi` are equally spaced then the weights `wi` can be computed
```
```suggestion
    x-coordinates `xi` are equally spaced, then the weights `wi` can be computed
```

mdhaber replied:
This was probably fine; I actually forgot that the inputs were named `xi` and `wi`.

j-bowhay replied:
I decided to change it anyway since it seemed odd for the notation to jump around

ev-br reviewed (approved):
Reads great and adds useful info!

A couple of minor optional suggestions below.
Consider also adding a note stressing that for $n$ data points it's still a polynomial of degree $n$ to clearly differentiate from piecewise interpolators. I know it's inferrable from  the current wording, so the suggestion's optional.
 Maybe drop the "global interpolation" term somewhere.

mdhaber commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,72 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+                  (-1)^i   & \text{otherwise}
+              \end{cases}.
+
+    Note that for large :math:`n` computing the weights explicitly (see examples) will
```
```suggestion
    Note that for large :math:`n`, computing the weights explicitly (see examples) will
```
Since it's come up a few times, I should probably include some justification, e.g. [here] (https://www.grammarly.com/blog/punctuation-capitalization/commas-after-introductory-clauses/).

mdhaber commented on scipy/interpolate/_polyint.py:
```diff
@@ -623,15 +668,34 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
     >>> plt.show()
+
+    Next, we show how using Chebyshev points of the second kind avoids the avoids the
+    Runge phenomenon. In this example we also compute the weights explicitly.
```
```suggestion
    Runge phenomenon. In this example, we also compute the weights explicitly.
```

mdhaber reviewed (commented):
This is cool. I forget - is finite precision in the Chebyshev points a practical consideration, or would it require absurd `n` to cause problems?

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +583,71 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+        w_i = \begin{cases}
+                  (-1)^i/2 & i = 0,n \\
+                  (-1)^i   & \text{otherwise}
+              \end{cases}.
```
Thanks, I was scratching my head over this

j-bowhay committed:
DOC: address review comments

[skip ci]

j-bowhay committed:
[docs only]

j-bowhay commented:
Thanks for all the reviews! I have hopefully addressed all the comments.

> This is cool. I forget - is finite precision in the Chebyshev points a practical consideration, or would it require absurd `n` to cause problems?

I think this should be fine, I ran the example with `n=1e5` with no issues (after that my computer ground to a halt). In fact, this idea forms the basis of chebfun, which has no issues approximating sufficiently nice functions. The question of large `n` is also covered in chpt 15 Approximation Theory and Approximation Practice by Nick Trefethen, if of interest.

mdhaber commented on scipy/interpolate/_polyint.py:
```diff
@@ -589,21 +587,71 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
-    Based on Berrut and Trefethen 2004, "Barycentric Lagrange Interpolation".
+    .. math::
+
+        w_i = (-1)^i n\choose i,
```
```suggestion
        w_i = (-1)^i {n \choose i},
```

mdhaber commented on scipy/interpolate/_polyint.py:
```diff
@@ -623,15 +671,34 @@ class BarycentricInterpolator(_Interpolator1DWithDerivatives):
+    >>> xx = np.linspace(-1, 1, 1000)
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(xx, f(xx), label="Original Function")
+    >>> ax.plot(xx, p_cheb(xx), label="Chebshev Points")
+    >>> ax.plot(xx, p_equi(xx), label="Equally Spaced Points")
```
```suggestion
    >>> ax.plot(xx, p_cheb(xx), "--", label="Chebshev Points")
    >>> ax.plot(xx, p_equi(xx), "--", label="Equally Spaced Points")
```

mdhaber committed:
Apply suggestions from code review

[skip ci]

mdhaber commented:
Thanks @j-bowhay!

mdhaber commented:
Curious why you chose a non-smooth function, though? Did you want to make a point about this?

![image](https://github.com/user-attachments/assets/a3e85aac-74c9-4fa4-b808-61078afcd61b)

j-bowhay commented:
> Curious why you chose a non-smooth function, though? Did you want to make a point about this?
> 
> ![image](https://private-user-images.githubusercontent.com/6570539/419627330-a3e85aac-74c9-4fa4-b808-61078afcd61b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyMDYzODcsIm5iZiI6MTc0MTIwNjA4NywicGF0aCI6Ii82NTcwNTM5LzQxOTYyNzMzMC1hM2U4NWFhYy03NGM5LTRmYTQtYjgwOC02MTA3OGFmY2Q2MWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDVUMjAyMTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDFiODg5YmRhNDAyYjY4Y2Y0ODJhNzdjNTEzYzZhY2ZhZjQ4MWU1MGJkNGNlYmMyMTcwY2VmZTMxNDgwYWRhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vIdaxOm-lHNzz9X7_ZIGlNR1OL0etf9Sj0dLsUZqXvk)

No especially profound reason; it was just an example in the paper!
